### ISSUE TITLE
Do not use CSD by default on Linux

### ISSUE THREAD

**Issue by ripefig** [Microsoft/vscode#65608]
CSD does not have good compatibility on Linux. VSCode's window has no shadows under KWin. VSCode's client-side decorations are inconsistent with both Gnome CSD and server-side decorations, both visually and functionally. 

Functional inconsitencies:

- Can't drag window from a widget area in the titlebar, unlike Gnome CSD. Drag area reduced to nothing.
- Doesn't follow right-click to maximize.
- Doesn't respect global titlebar button configurations, on any desktop.

Visual inconsistencies are obvious.

CSD could be default if it was at least consistent with Gnome CSD. Otherwise, I believe this feature should remain disabled by default because the only benefit is a few saved pixels but the downsides are numerous.

**Comment by francoism90** [comment]
@ripefig How to disable CSD?

**Comment by zhangboyang** [comment]
> @ripefig How to disable CSD?

In "Settings", search `window.titleBarStyle` and set it to `native`.

**Comment by sbatten** [comment]
keeping this one open as a way to guage feedback on this issue. the custom titlebar provides various accessibility improvements over the electron implementation. as always, this can be disabled with 
`window.titleBarStyle` => `native`

**Comment by wildcart** [comment]
It's not only the style and order of the buttons, as mentioned in #65241, but also their position: My close button is on the left while the rest is on the right (and it's showing more than min/max/close) and when the window is maximised, the title bar is hidden.

Additionally, my menu is placed in a global menu at the top of the screen (hidden by default) which is also broken when the custom title bar is enabled. So I was surprised to see the VS Code menu again, after updating the package. 

Screenshot showing Code 1.30 running on KDE Plasma 5.14 using the custom titlebar
<img width="1276" alt="code-1 30-custom_titlebar-plasma-5 14" src="https://user-images.githubusercontent.com/19203041/52203838-acae5600-2872-11e9-802d-dfa2081cbd74.png">

This can be fixed by using the native title bar, after which 1.30 looks and behaves like 1.23 (the version I ran previously).

Screenshot showing Code 1.23 running on KDE Plasma 5.14 (global manu hidden)
<img width="1278" alt="code-1 23-plasma-5 14" src="https://user-images.githubusercontent.com/19203041/52203849-b768eb00-2872-11e9-8e44-77d1b790b5a2.png">

Screenshot showing Code 1.23 running on KDE Plasma 5.14 (global menu shown)
<img width="1277" alt="code-1 23-plasma-5 14-menu" src="https://user-images.githubusercontent.com/19203041/52203859-bdf76280-2872-11e9-970f-212956942887.png">



Some general words about window decorations (because I think it's important especially for Linux users):
It's the job of the window manager, and not that of the application, to render the window decoration because it ensure a coherent UI/UX across all applications. I know that Windows has a very different take on it <opinionated> and look at the mess it creates. 'Every single' software (company) uses its own UI/UX offering different functionality. Even Microsoft programs follow different guidelines (at least 3 as of Windows 10, with VS Code's custom title bar adding one more).</opinionated> But on Linux (and Mac) the window decoration is almost always rendered by the window manager (chromium being an exception, which has an option to not use native title bars - in my distro on by default). This is import to note because each window manager may offer more functionality than simply min/max/closing a window, most importantly moving windows to different virtual desktops, keep-on-top, keep-below, sticky and others(more details in #65241). Letting the application render the title bar makes it impossible to access this functionality which creates a bad user experience, because users don't understand why a certain function they are used to is not working in a certain application while it works fine in 'all' others. Some window managers may offer 'other' ways to access the title bar right-click-menu, KWin has the ALT-F3 shortcut, but this is normally not known by everyday-[Linux-]users, hence the functionality is hidden/inaccessible!

In summary, the custom title bar breaks a lot of feature on Linux, the benefits compared to its drawbacks for Linux users is, well, zero.


Thanks for considering
  Christian.

**Comment by sbatten** [comment]
As has been stated, the custom title bar and menus provide a more cohesive experience within the application regarding theming along with the primary focus of custom menus, accessibility improvements. I do understand that this is not a priority for all Linux users and this is why the setting to revert will remain an option.

Arguing that we should support the various different things that DE can contribute on Linux as we have matched Windows/macOS is not a fair comparison as Linux is highly customizable. In this regard, we have once again, opted to allow the user to revert to the setting that allows them to keep all DE customizations that they prefer. Some customizations like having the the window controls on the left side are a bit simpler which is why that issue has remained open for consideration. Supporting different icon styles, colors, heights, shadows, etc are all likely beyond the scope of VS Code.

**Comment by wildcart** [comment]
> Arguing that we should support the various different things that DE can contribute on Linux as we have matched Windows/macOS is not a fair comparison as Linux is highly customizable.

Thank you for making the point for us: Simply don't! 

It may create a "more cohesive experience within the application" but it totally breaks the "cohesive experience" of a user across all applications on the platform the application runs. Do you really thing that each application shipping its own "cohesive experience" is good for overall user experience?

**Comment by zachbr** [comment]
Ultimately @wildcart's reply is the exact point I was trying to make.

> In this regard, we have once again, opted to allow the user to revert to the setting that allows them to keep all DE customizations that they prefer.

Breaking desktop cohesion by default and then offering the user a choice to revert it is backwards. I would absolutely again draw you to the comparison to Windows or macOS applications. Many of them do not fully integrate with Windows or macOS as a platform, but they don't ship their own visually distinct title bar or window frame or other major UI component, and those few that do must either take special care to fit the environment or are rightly ostracized for it.

> As has been stated, the custom title bar and menus provide a more cohesive experience within the application

So it's fine to break desktop cohesion if it means the application appears to be more cohesive? That doesn't sound well rooted in any platform or application design schema I've ever heard of. I'd be willing to bet there are guidelines that state just the opposite somewhere on Microsoft's site advising Windows application developers to avoid it.

> theming along with the primary focus of custom menus, accessibility improvements

If this is the crutch of the argument perhaps we should dive deeper into custom menus and enumerate the accessibility improvements so that it is possible for the vscode linux community, and even the larger vscode community to have a more informed and encompassing discussion.

**Comment by ripefig** [comment]
I am seeing lots of awful CSD from Electron apps popping up. Electron needs to provide a way of constructing CSDs that 100% visually and functionally identical to those used by GTK. Before that happens, CSDs shouldn't be used at all.

But then this leaves

**Comment by ripefig** [comment]
The problem with CSDs on Linux is fundamental. It took Mozilla a whole year just to create a CSD that looks at home in a Gnome environment (of course you can forget KDE).

The only way to really solve this is DWD, a project that KDE started several years ago and then abandoned: https://kver.wordpress.com/2014/10/25/presenting-dwd-a-candidate-for-kde-window-decorations/

**Comment by sbatten** [comment]
I appreciate all the passionate feedback in this thread. After cosidering the points here and the value of the custom title bar on Linux, we have decided to revert the default for Linux. A short blurb about this change can be found in our [Linux docs page](https://code.visualstudio.com/docs/setup/linux#_using-the-custom-title-bar). 

fixed by #68640

**Comment by soncodi** [comment]
FWIW, I prefer the `custom` title bar to the `native` look-and-feel. It also goes well with my DE's dark theme. I don't care which is the default, but please don't remove the custom option.

**Comment by m-thorsen** [comment]
Could it be an idea to instead revisit a GTK dark mode toggle for better Linux integration? ( Ref: #11979 )  
I'm sure this could be extended to other OSes as well seeing that dark mode is a thing in both macos and windows now.

There's an extension for this ( https://github.com/fkrull/vscode-gtk-dark-titlebar ), but that extension doesn't disable itself on other OSes, causing issues when syncing settings across platforms.

**Comment by sbatten** [comment]
@m-thorsen One of the issues I saw in the linked thread and is also mentioned in the extension README is that the menu doesn't change theme with GTK. Is this fixed now with some of the changes that went into electron around menu theming?

**Comment by m-thorsen** [comment]
@sbatten No, the method suggested in that thread does not change the menu theme.

There are a few issues filed with Electron on this issue - to link a few: 
https://github.com/electron/electron/issues/9681  
https://github.com/electron/electron/issues/13381  
https://github.com/electron/electron/pull/15878  

(Disclaimer: I haven't read through all the issue comments)

Either way I think a dark title bar would be a good starting point - as it at least blends in when the menu bar is hidden.
